### ISSUE TITLE
PLC0208 false positive when a set contains equal values of different types

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#14049]
The fix for [`iteration-over-set` (PLC0208)](https://docs.astral.sh/ruff/rules/iteration-over-set/) has false positives which change behavior. It intentionally ignores sets that contain duplicates, but when detecting duplicates it assumes that literals of different types are never equal. That is not true; for example, `False == 0`, so `{False, 0}` contains a duplicate and is actually equal to `{False}`.

```console
$ ruff --version
ruff 0.7.2

$ cat plc0208.py
for x in {False, 0, 0.0, 0j, True, 1, 1.0}:
    print(x)

$ python plc0208.py 
False
True

$ ruff check --isolated --select PLC0208 plc0208.py --fix
Found 1 error (1 fixed, 0 remaining).

$ cat plc0208.py
for x in (False, 0, 0.0, 0j, True, 1, 1.0):
    print(x)

$ python plc0208.py
False
0
0.0
0j
True
1
1.0
```

**Comment by MichaReiser** [comment]
Related to https://github.com/astral-sh/ruff/issues/12772

**Comment by charliermarsh** [comment]
We might want some sort of analogue to `ComparableExpr` that implements these semantics.
